optimisation/design/Third_Party_API_Design.md

```diff
@@ -0,0 +1,595 @@
+For example, if the C++ library `samplelib` defines a `Crystal` object with properties `BMatrix` and `Orientation` and method `Rotate`,
+the properties may be obtained in Matlab by `xtal = clib.samplelib.Crystal(abc, alf); B = xtal.BMatrix;`
+and `rotated_xtal = xtal.Rotate(alpha, beta, gamma); new_orientation = rotated_xtal.Orientation`. 
+
+Whilst the `clib` approach allows C++ classes to be more directly accessed in Matlab,
+it is not necessary for our purposes, which requires only calling a single function.
+Furthermore, using `clib` requires a compiled wrapper library to be generated using `clibgen`,
+which should be done for each user generated function.
+
+*For these reasons, we propose to use the* `loadlibrary` *and* `calllib` *method instead, with a predefined header file.*
+This header file (which must be a file - this is a limitation of `loadlibrary`) may be generated on-the-fly
+in order to accommodate custom user function names.
+Also, if we are to accommodate method 3 for inferring input arguments [noted above](#external_interfaces)
+then this header file must be generated dynamically.
+
+The actual function signatures and API are discussed for each language separately in the sections below, 
+but we should note here a limitation with using `loadlibrary` (and using dynamically loaded libraries in general).
+This is because the C-language function may only accept pointers to arrays,
+and because, as noted [in the appendix](app_nativememory) the main Matlab code and library codes have different memory heaps,
+so they must each be responsible for allocating *and deallocating* their own memory.
+That is, a memory error will occur if Matlab tries to deallocate a memory block allocated by the library and vice versa.
+To avoid this, *we propose to not allow user functions to allocate their own memory for the results array*.
+That is, Matlab will allocate an array for the results and pass the pointer to this to the user model function.
+The user model function must then fill this array with the result of the calculations,
+rather than using their own allocated array.
+
+Finally we note that the compiled library interface deal *exclusively* with real-valued variables.
+The user defined model function is supplied with real input coordinates ![\mathbb{Q}] and is expected to return real ![S(\mathbb{Q})].
+It should thus not return complex values, even if the calculations result in complex values.
+In these cases it should either raise an error or a warning.
+
+
+# User defined model function signatures
+
+## <a name="matlab_functions"></a> Matlab
+
+```matlab
+function user_model_sqw(qh, qk, ql, en, parameters, varargin)
+function user_model_dsp(qh, qk, ql, parameters, varargin)
+function user_model_pow(modQ, en, parameters, varargin)
+function user_model_1d(en, parameters, varargin)
+function user_model_0d(parameters, varargin)
+```
+
+This follows the convention established by the current implementation of `multifit`.
```

Comment:
For Matlab and Python user functions, probably it doesn't matter either way. For the compiled library there might be an ambiguity on the data layout - is the Matlab 2D array passed column or row-major? Newer Matlab supports C-style row-major matrices but the ones I've tested all return a column-major array.

Also, unless we add another `int` to the function signature (in addition to `n_elem`) to determine the size of the input array, we would need to trust that the users will handle the input array size correctly. With multiple 1D input arrays we can use the runtime signatures checking in `calllib` to ensure that the correct library function is called.

So, on balance I would prefer to stick with multiple 1D arrays, unless it would be much more performant for Horace to pass a 2-D array.